### PR TITLE
Drop FutureNCCL in favor of vanilla CUDAFuture

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #48790 Add support for async callbacks in ivalue::Future
* #48840 Remove DataPtr extractor from CUDAFuture
* **#49014 Drop FutureNCCL in favor of vanilla CUDAFuture**

We extracted a generic and reusable CUDAFuture class from FutureNCCL, but we had left FutureNCCL around, as a subclass of CUDAFuture, in order to deal with some peculiarity of ProcessGroupNCCL, namely that the future would be completed right away when constructed and that its CUDA events would be _shared_ with the ones of the WorkNCCL. This required some "hacks" in CUDAFuture itself (protected members, fields wrapped in shared_ptrs, ...).

My understanding is that creating CUDA events is a rather cheap operation. That would mean that we could afford to record _twice_ the events after each NCCL call, once for the WorkNCCL and once for the future. By doing so, we can use the CUDAFuture class directly and revert all its hacks.

Differential Revision: [D25355272](https://our.internmc.facebook.com/intern/diff/D25355272/)

### CODE REVIEW

**Review by mrshenli — CHANGES_REQUESTED** [review]
test failures look relevant on this one

File: torch/lib/c10d/ProcessGroupNCCL.hpp [hunk 1]
```diff
@@ -190,62 +194,6 @@ class ProcessGroupNCCL : public ProcessGroup {
...
-  // outputs' devices. Its value is NCCL collective's outputs.
-  //
-  // If created by FutureNCCL's then callback, its value becomes the value of
-  // callback() and its cudaEvents will record the NCCL stream that runs that
-  // callback. Before invoking the callback, FutureNCCL will synchronize its
-  // own cudaEvents with the stream that runs the callback. This design
-  // enables synchronizing the appropriate streams and avoids stalling PyTorch's
-  // default stream while running the callback. In case of multiple then
-  // callbacks, each will be executed on its own fresh stream.
-  struct FutureNCCL : at::cuda::CUDAFuture {
```
Comment (mrshenli) [inline torch/lib/c10d/ProcessGroupNCCL.hpp:211]: nice!

Comment (mrshenli) [reply torch/lib/c10d/ProcessGroupNCCL.hpp:211]: `FutureNCCL` is also mentioned in our docs. Let update those too:

https://github.com/pytorch/pytorch/blob/dee82ef3ea1ed2cae4e891f071adff03cb4a6e0d/torch/csrc/distributed/c10d/init.cpp#L1140-L1144

**Review by mrshenli — COMMENTED** [review]
Rest of the change LGTM!

**Comment by lw** [comment]
I think I have fixed the build failures

**Review by mrshenli — APPROVED** [review]
LGTM!

**Pull request closed without merging**